feat: add helper-functions.bash collection w/ flox_portgrab()

limeytexan opened the pull request:
## Proposed Changes

This patch adds a new place for adding bash helper functions to be used within the on-activate hook, and uses it to add a new flox_portgrab() function. This function is useful in the service management area for grabbing an ephemeral port that is guaranteed to be free and available for use.

Example usage:

export SERVER_PORT=$(flox_portgrab)

## Release Notes

* added new `flox_portgrab()` function for use within `hooks.on-activate` activation section of the manifest

limeytexan committed:
feat: add helper-functions.bash collection w/ flox_portgrab()

This patch adds a new place for bash helper functions to be used within
the on-activate hook, and uses it to add a new flox_portgrab() function.
This function is useful in the service management area for grabbing an
ephemeral port that is guaranteed to be free and available for use.

Example usage:

SERVER_PORT=$(flox_portgrab)

rossturk commented:
I like the idea of having a helper function file available for producers that carries out common tasks. I might contribute functions here that:
* create and prepare python venvs
* read a specific env variable from a file

stealthybox commented:
Ooh I love this -- so much room for useful helper funcs.
Maybe this is already detailed somewhere, but I wonder how this could work with other shells?

1. Grabbing a random port makes a lot of sense for configuring some API server you're working on to connect to a database. 

2. I can anticipate it also feeling potentially annoying if the port number is super pertinent to the dev/practitioner's workflow when using the env.
I don't know how much I'd appreciate the port number changing on me every time I activate my env.

Maybe it doesn't matter if there are other niceties like the env printing a URL or automatically launching mysqlWorkbench/Jupyter/etc properly configured, but it's also likely that people would have to copy the ip:port number and input it manually into another tool.

3. It could be useful to put in a range or preference list of ports, so that the port that comes back is canonically related to the app/lang-ecosystem.
Ex: if I'm using NodeJS I might want a port within the 3000..3100 range with a fallback.

3b. the smaller the port preference list, the greater the change of races, so maybe supporting that requires writing/reading from a sqlitedb or something.

4. This is quickest to implement and iterate on as a shell func, and the practitioner can easily copy/fork the logic.
The tradeoff is we have to support multiple shells / OS's / hardware.
Maybe it's more portable across those dimensions if we build the helper logic into a flox subcommand or other helper binary?

stealthybox commented:
Other helper funcs that could be useful:

`flox_ipv4grab()`
Given a specific port number, find me an unused 127.0.0.0/8 IPv4 addr to bind to

`flox_ipv6grab()`
Given a specific port number, find me an unused local IPv6 addr to bind to

`flox_ensure_line( file, line )`
Make sure file contains a line

`flox_wait_for_tcp( timeout )`
`flox_wait_for_tls( timeout )`

The docs-site/man page for these should probably be next to a bunch of examples for:
- creating/reading/editing structured config files like json, yaml, toml/ini, xml, cue, nickel, pkl using their associated command line tools + links to those packages on FloxHub
- minting ca's/certs with openssl / cfssl / mkcert

mkenigs commented:
Discussed and we'll probably merge this once we're ready for it (may need to move the man page changes to the services folder), but we're not quite ready for it with where we are with services

dcarley committed:
test(activate): Test for flox_portgrab

Add a new test for the `flox_portgrab` helper within `on-activate` by
verifying that we can get a 5-digit port number as an env var.

dcarley committed:
fix(activate): Fix flox_portgrab args for nc

The version of netcat that we get takes the port as a positional
argument. The test was failing because it was unable to match the
expected output against the usage output:

    % /nix/store/aj3i0mqh3cfp2yr16kn034s9c6p1mzbi-libressl-3.8.2-nc/bin/nc 0.0.0.0 -l 0 -n -v
    usage: nc [-46cDdFhklNnrStUuvz] [-C certfile] [-e name] [-H hash] [-I length]
	     [-i interval] [-K keyfile] [-M ttl] [-m minttl] [-O length]
	     [-o staplefile] [-P proxy_username] [-p source_port] [-R CAfile]
	     [-s sourceaddr] [-T keyword] [-V rtable] [-W recvlimit] [-w timeout]
	     [-X proxy_protocol] [-x proxy_address[:port]] [-Z peercertfile]
	     [destination] [port]

This was a bit of a pain to debug because:

- changes to the scripts require a rebuild of the environment
- the errors can swallowed by the redirect
- debugging via `flox activate -v` is quite limited

Which, combined with `on-activate` being run multiple times, makes me
feel like we shouldn't release this helper just yet.

dcarley commented:
I looked at pulling this in for https://github.com/flox/flox/pull/1801 but my feeling is that we shouldn't expose it to users until we have a way to deal with multiple activations getting different values, which is problematic for a client discovering the server. We could provide some extra sugar to store named values in `$FLOX_ENV_CACHE` but it locks us into an interface and is more work than I was hoping for the current ticket. In the meantime though I've pushed a couple of commits which add a test and fixes a small argument issue.

limeytexan commented:
There have been several occasions for which this functionality would have been useful, and we've needed to duplicate the ephemeral port allocation function into the environments that needed it. This PR clearly needs a manual rebase, but that would be relatively quick to do. Flagging it for discussion amongst the product team to hopefully get this over the line. @jbayer @djsauble @Kaysahni

billlevine commented:
Tracked in Linear: **DEV-51** — "Provide a way for services to request an ephemeral port within the on-activate hook" (Developer Workflows team).

https://linear.app/floxdotdev/issue/DEV-51/provide-a-way-for-services-to-request-an-ephemeral-port-within-the-on

---
*Via Forge (interactive) • 6c110fd1*
